tools/inc/stljobs.h

```diff
@@ -0,0 +1,606 @@
+            PIPE_TYPE_BYTE | PIPE_READMODE_BYTE | PIPE_WAIT | PIPE_REJECT_REMOTE_CLIENTS, 1, 0, kernelBufferSize, 0,
+            nullptr);
+
+        if (!SetFileCompletionNotificationModes(
+                readHandle.get(), FILE_SKIP_COMPLETION_PORT_ON_SUCCESS | FILE_SKIP_SET_EVENT_ON_HANDLE)) {
+            api_failure("SetFileCompletionNotificationModes");
+        }
+
+        SECURITY_ATTRIBUTES inheritSa;
+        inheritSa.nLength              = sizeof(inheritSa);
+        inheritSa.lpSecurityDescriptor = nullptr;
+        inheritSa.bInheritHandle       = TRUE;
+        writeHandle = create_file(pipeNameBuffer, GENERIC_WRITE | FILE_READ_ATTRIBUTES, 0, &inheritSa, OPEN_EXISTING,
+            FILE_FLAG_OVERLAPPED, NULL);
+
+        readIo = tp_io{std::move(readHandle), callback, this, nullptr};
+
+        start();
+    }
+
+    ~output_collecting_pipe() {
+        if (readIo) {
+            if (running.load()) {
+                if (!CancelIoEx(readIo.get_file(), &overlapped)) {
+                    api_failure("CancelIoEx");
+                }
+            }
+
+            readIo.close();
+        }
+    }
+
+    output_collecting_pipe(const output_collecting_pipe&) = delete;
+    output_collecting_pipe& operator=(const output_collecting_pipe&) = delete;
+
+    void start() {
+        const auto oldRunning = running.exchange(true);
+        assert(!oldRunning);
+        (void) oldRunning;
+        read_some();
+    }
+
+    void stop() {
+        if (running.exchange(false)) {
+            if (!CancelIoEx(readIo.get_file(), &overlapped)) {
+                api_failure("CancelIoEx");
+            }
+
+            readIo.wait(false);
+        }
+    }
+
+    [[nodiscard]] std::string extract() {
+        stop();
+        std::string result(targetBuffer.data(), validTill);
+        start();
+        return result;
+    }
+
+    [[nodiscard]] std::string extract_and_reset() {
+        stop();
+        auto first = targetBuffer.data();
+        auto last  = first + validTill;
+        first      = std::find_if_not(first, last, is_space);
+        last       = std::find_if_not(std::reverse_iterator(last), std::reverse_iterator(first), is_space).base();
+        std::string result(first, static_cast<size_t>(last - first));
+        validTill = 0;
+        start();
+        return result;
+    }
+
+    [[nodiscard]] const std::string& get_buffer() const noexcept {
+        assert(!running.load());
+        return targetBuffer;
+    }
+
+    [[nodiscard]] HANDLE get_write_pipe() noexcept {
+        return writeHandle.get();
+    }
+
+private:
+    static void __stdcall callback(
+        PTP_CALLBACK_INSTANCE, void* thisRaw, void*, ULONG ioResult, ULONG_PTR bytes, PTP_IO) noexcept {
+        switch (ioResult) {
+        case ERROR_SUCCESS:
+        case ERROR_OPERATION_ABORTED:
+            break;
+        default:
+            api_failure("StartThreadpoolIo callback", ioResult);

```

Comment:
Nitpick: Failure here indicates a completion of ReadFile API with error, not a failure of StartThreadpoolIo. A future reader (of the code or logs) may be confused about which API failed.

Also, we are throwing here on a threadpool thread that does not have a try-catch and won't nice print anything. I am wondering, if we would want to log something in this case so that we are not silent when it happens. For example, we can log in the `api_failure` itself.